charmdir.go

```diff
@@ -123,6 +139,33 @@ func ReadCharmDir(path string) (dir *CharmDir, err error) {
 	return dir, nil
 }
 
+// setupIgnoreRules parses the contents of the charm's .jujuignore file and
+// compiles a set of rules that are used to decide which files should be
+// archived.
+func (dir *CharmDir) setupIgnoreRules() error {
+	pathToJujuignore := dir.join(".jujuignore")
+	if _, err := os.Stat(pathToJujuignore); err == nil {
+		file, err := os.Open(dir.join(".jujuignore"))
+		if err != nil {
+			return err
+		}
+		defer func() { _ = file.Close() }()
+
+		dir.ignoreRules, err = newIgnoreRuleset(file)
+		if err != nil {
+			return errors.Annotate(err, ".jujuignore")
+		}
+	}
+
+	defaultRules, err := newIgnoreRuleset(strings.NewReader(defaultJujuIgnore))
+	if err != nil {
+		return err
+	}
+
+	dir.ignoreRules = append(dir.ignoreRules, defaultRules...)
```

Comment:
Is this right, where `defaultRules...` will overwrite the charm `ignoreRules`?

So if you have `foo` in default `.jujuignore` and then `!foo` in charm `.jujuignore` then it won't be included. Even if that's not possible, we should probably be consistent and add `defaultRules` first to be consistent?

Reply:
You have a point there. Makes more sense to do it the other way round. 

An alternative we may want to consider is to use the charm's .jujuignore file if one is provided OR use the defaults as a fallback. This way there will be no unexpected surprises for people writing .jujuignore files. Thoughts?

Reply:
I think it should be safe to just flip it the other way around (defaults first), mainly because I think it should be perfectly safe for them to do `!.git` if they require. Whilst also providing some very safe defaults for most users and it will help with cutting down on files we have to manage in general, as apposed to not providing one.

Reply:
Also `.git` folder is entirely dangerous given that secrets accidentally committed and then removed from git history could still be extracted. So having sane defaults seems like a pure win.